applications/graph/MeshGraphNet/GNN.py

```diff
@@ -0,0 +1,114 @@
+import lbann
+from GNNComponents import MLP, GraphProcessor
+
+
+def input_data_splitter(input_layer,
+                        num_nodes,
+                        num_edges,
+                        in_dim_node,
+                        in_dim_edge,
+                        out_dim):
+  """ Takes a flattened sample from the Python DataReader and slices
+      them according to the graph attributes.
+  """
+
+  split_indices = []
+  start_index = 0
+  node_feature_size = num_nodes * in_dim_node
+  edge_feature_size = num_edges * in_dim_edge
+  out_feature_size =  num_nodes * out_dim
+
+  split_indices.append(start_index)
+  split_indices.append(split_indices[-1] + node_feature_size)
+  split_indices.append(split_indices[-1] + edge_feature_size)
+  split_indices.append(split_indices[-1] + num_edges)
+  split_indices.append(split_indices[-1] + num_edges)
+  split_indices.append(split_indices[-1] + out_feature_size)
+
+  sliced_input = lbann.Slice(input_layer, axis=0, slice_points=split_indices)
+
+  node_features = lbann.Reshape(lbann.Identity(sliced_input),
+                                dims=[num_nodes, in_dim_node])
+  edge_features = lbann.Reshape(lbann.Identity(sliced_input),
+                                dims=[num_edges, in_dim_edge])
+  source_node_indices = lbann.Reshape(lbann.Identity(sliced_input), dims=[num_edges])
+  target_node_indices = lbann.Reshape(lbann.Identity(sliced_input), dims=[num_edges])
+
+  out_features = lbann.Reshape(lbann.Identity(sliced_input),
+                               dims=[num_nodes, out_dim])
+
+  return node_features, edge_features, source_node_indices, \
+    target_node_indices, out_features
+
+
+def LBANN_GNN_Model(num_nodes, num_edges,
+                    in_dim_node, in_dim_edge, out_dim,
+                    out_dim_node=128, out_dim_edge=128,
+                    hidden_dim_node=128, hidden_dim_edge=128,
+                    hidden_layers_node=2, hidden_layers_edge=2,
+                    mp_iterations=15,
+                    hidden_dim_processor_node=128, hidden_dim_processor_edge=128, 
+                    hidden_layers_processor_node=2, hidden_layers_processor_edge=2,
+                    norm_type=lbann.LayerNorm,
+                    hidden_dim_decoder=128, hidden_layers_decoder=2,
+                    num_epochs=10):
+  
+  # Set up model modules and associated weights 
+
+  node_encoder  = MLP(in_dim=in_dim_node, out_dim=out_dim_node,
+                      hidden_dim=hidden_dim_node, hidden_layers=hidden_layers_node,
+                      norm_type=norm_type, name="graph_input_node_encoder")
+
+  edge_encoder  = MLP(in_dim=in_dim_edge, out_dim=out_dim_edge,
+                      hidden_dim=hidden_dim_edge, hidden_layers=hidden_layers_edge,
+                      norm_type=norm_type, name="graph_input_edge_encoder")
+
+  graph_processor = GraphProcessor(num_nodes=num_nodes,
```

Comment:
Maybe it's worth noting in the code that there is no distinction between mesh edges and world edges in this implementation, nor adaptive remeshing.

Reply:
Added some documentation in the code and in the read me